README.md
# history-service

Service that provides call log and conversation history

## Getting Started


History service provide the database and an API to store/retrieve the call log (used by dialer-app ) and the sms/mms history ( used by messaging-app ).

See as well telepathy-ofono for incoming message events.

Database location: ~.local/share/history-service/history.sqlite


### Installing

You can use [crossbuilder](http://docs.ubports.com/en/latest/systemdev/testing-locally.html#cross-building-with-crossbuilder)
( you may need to add manually dh-translations, e.g `crossbuilder inst-foreign dh-translations)

```
crossbuilder
```

## Running the tests

Run tests within the container `crossbuilder shell` and find the generated tests, currently on `history-service/obj-..../tests/`


## Contributing

Please read [CONTRIBUTING.md](http://docs.ubports.com/en/latest/systemdev/testing-locally.html).


## License

GPL v3.0 - see the [COPYING](COPYING) file for details



debian/changelog
history-service (0.3~0ubports4) xenial; urgency=medium

  * Revert use of sent date instead of the received one

 -- Lionel Duboeuf <lduboeuf@ouvaton.org>  Tue, 09 Jul 2020 20:44:04 +0200

history-service (0.3~0ubports3) xenial; urgency=medium

  * Allow querying history db with NotEquals match flag

 -- Lionel Duboeuf <lduboeuf@ouvaton.org>  Tue, 09 Jul 2020 20:44:04 +0200

history-service (0.3~0ubports2) xenial; urgency=medium

  * Add MessageStatusDraft Enum

 -- Lionel Duboeuf <lduboeuf@ouvaton.org>  Tue, 31 Mar 2020 20:44:04 +0200

history-service (0.3~0ubports1) xenial; urgency=medium

  * debian/control:
    Allow building with crossbuilder
  * Update to new QtPim

 -- Alberto Mardegan <mardy@users.sourceforge.net>  Tue, 18 Sep 2018 20:44:04 +0300

history-service (0.2+0ubports2) xenial; urgency=medium

  * No change rebuild against new QtPim.

 -- Alberto Mardegan <mardy@users.sourceforge.net>  Wed, 25 Jul 2018 21:46:57 +0300

history-service (0.2+0ubports1) xenial; urgency=medium

  * Adapt to support VOIP accounts.
  * Improve the notifications of participants changing
  * Only start saving information events about contacts joining and leaving after the self contact is in the local list of participants. 
  * Improve Roles management performance by caching the retrieved data.
  * Mark entire conversations as read.
  * Allow pass multiple fields on sort clause.
  * Reduce the dbus traffic when marking messages and threads as read.
  * Use a QLockFile to ensure there will be only one instance of the daemon per user. As we now delay the registration on dbus, sometimes we ended up having two instances of the daeon running (because of dbus activation). This change makes sure that won't happen.
  * Do not load the participants from threads automatically. If the client really needs it, it can use the newly added API to fetch the participants.
  * Make it possible to debug sqlite commands.

 -- Marius Gripsgard <marius@ubports.com>  Tue, 06 Mar 2018 08:18:08 +0100

history-service (0.2+ubports) xenial; urgency=medium

  * Imported to UBports

 -- UBports auto importer <infra@ubports.com>  Thu, 11 Jan 2018 10:22:35 +0000

history-service (0.1+17.04.20161130-0ubuntu1) zesty; urgency=medium

  [ Gustavo Pichorim Boiko ]
  * Improve group chat support.

 -- Tiago Salem Herrmann <tiago.herrmann@canonical.com>  Wed, 30 Nov 2016 15:14:35 +0000

history-service (0.1+17.04.20161124.1-0ubuntu1) zesty; urgency=medium

  * Append "SNAP" prefix into the path to the plugins dir if none is
    set.

 -- Renato Araujo Oliveira Filho <renato.filho@canonical.com>  Thu, 24 Nov 2016 18:54:43 +0000

history-service (0.1+16.10.20160909.1-0ubuntu1) yakkety; urgency=medium

  * Temporarily disable DaemonTest until we get QtDBus fixed.

 -- Gustavo Pichorim Boiko <gustavo.boiko@canonical.com>  Fri, 09 Sep 2016 14:53:42 +0000

history-service (0.1+16.04.20160425-0ubuntu1) xenial; urgency=medium

  * Make sure Telepathy types are always initialized. (LP: #1563050)

 -- Gustavo Pichorim Boiko <gustavo.boiko@canonical.com>  Mon, 25 Apr 2016 21:03:28 +0000

history-service (0.1+16.04.20160104-0ubuntu1) xenial; urgency=medium

  [ Tiago Salem Herrmann ]
  * Do proper dbus demarshalling of detailProperties. (LP: #1510655)

 -- Bill Filler <ci-train-bot@canonical.com>  Mon, 04 Jan 2016 20:22:09 +0000

history-service (0.1+16.04.20151207-0ubuntu1) xenial; urgency=medium

  [ CI Train Bot ]
  * New rebuild forced.

  [ Tiago Salem Herrmann ]
  * Allow removing empty threads (LP: #1517602)
  * Fix build with new qt on xenial-proposed.
  * Mark messages as read on onMessageRead(). (LP: #1373059)

 -- Tiago Salem Herrmann <tiago.herrmann@canonical.com>  Mon, 07 Dec 2015 17:55:02 +0000

history-service (0.1+16.04.20151103-0ubuntu1) xenial; urgency=medium

  [ Tiago Salem Herrmann ]
  * Enable cross compiling.

 -- Gustavo Pichorim Boiko <gustavo.boiko@canonical.com>  Tue, 03 Nov 2015 14:26:16 +0000

history-service (0.1+15.10.20151020.1-0ubuntu1) wily; urgency=medium

  [ Gustavo Pichorim Boiko ]
  * Make it possible to use the custom functions defined by the sqlite
    plugin in the schema files themselves.
  * Move the classes from historyprivate static library to
    libhistoryservice itself and just keep their header files private.
  * Optimize the thread matching by avoiding some phone number
    formatting and validating. (LP: #1488988)
  * Request contact information for all known participants on history-
    daemon initialization, and use this cached information on the
    models.
  * Store threads' timestamps in UTC in the cache, and convert only when
    returning to the client, so that it has the most up-to-date local
    time.

  [ Tiago Salem Herrmann ]
  * Initial cache implementation for grouped conversations.
  * Set max count hint Don't use fallback to tel to avoid queries to the
    address-book-service with null filters

 -- CI Train Bot <ci-train-bot@canonical.com>  Tue, 20 Oct 2015 20:22:05 +0000

history-service (0.1+15.10.20150828.1-0ubuntu1) wily; urgency=medium

  [ Gustavo Pichorim Boiko ]
  * Do not override the build type. Also remove a block of unused code
    from debian/rules.
  * Fix flags.
  * Prevent the history-daemon from crashing in the event of an ended
    call not having any remote member. (LP: #1458990)

  [ Tiago Salem Herrmann ]
  * Correctly set mCanFetchMore in the HistoryGroupedThreadsModel.
  * Generate eventId when there is none Set status to "Accepted" if the
    channel does not support delivery reports
  * Manually add self id to participants list if needed. (LP: #1486187)

 -- CI Train Bot <ci-train-bot@canonical.com>  Fri, 28 Aug 2015 10:16:13 +0000

history-service (0.1+15.10.20150819.2-0ubuntu1) wily; urgency=medium

  [ Gustavo Pichorim Boiko ]
  * Fix tests on vivid.

  [ Tiago Salem Herrmann ]
  * Wait telepathy to be ready before querying address-book-service.
    (LP: #1485005)

 -- CI Train Bot <ci-train-bot@canonical.com>  Wed, 19 Aug 2015 22:29:24 +0000


src/types.h
/*
 * Copyright (C) 2013-2016 Canonical, Ltd.
 *
 * Authors:
 *  Gustavo Pichorim Boiko <gustavo.boiko@canonical.com>
 *
 * This file is part of history-service.
 *
 * history-service is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 3.
 *
 * history-service is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef TYPES_H
#define TYPES_H

#include <QSharedPointer>
#include <QFlags>

#define DefineSharedPointer(type) class type; typedef QSharedPointer<type> type##Ptr; typedef QWeakPointer<type> type##WeakPtr; typedef QList<type##Ptr> type##s;

namespace History
{

DefineSharedPointer(EventView)
DefineSharedPointer(Plugin)
DefineSharedPointer(ThreadView)

// enums
enum EventType {
    EventTypeText,
    EventTypeVoice,
    EventTypeNull
};

enum FilterType {
    FilterTypeStandard,
    FilterTypeIntersection,
    FilterTypeUnion
};

enum MatchFlag {
    MatchCaseSensitive = 0x01,
    MatchCaseInsensitive = 0x02,
    MatchContains = 0x04,
    MatchPhoneNumber = 0x08
};

Q_DECLARE_FLAGS(MatchFlags, MatchFlag)

enum MessageStatus
{
    MessageStatusUnknown,
    MessageStatusDelivered,
    MessageStatusTemporarilyFailed,
    MessageStatusPermanentlyFailed,
    MessageStatusAccepted,
    MessageStatusRead,
    MessageStatusDeleted,
    MessageStatusPending, // pending attachment download
    MessageStatusDraft
};

enum MessageType
{
    MessageTypeText = 0,
    MessageTypeMultiPart = 1,
    MessageTypeInformation = 2
};

enum InformationType
{
    InformationTypeNone = 0,
    InformationTypeSimChange = 1,
    InformationTypeText = 2,
    InformationTypeSelfJoined = 3,
    InformationTypeJoined = 4,
    InformationTypeTitleChanged = 5,
    InformationTypeInvitationSent = 6,
    InformationTypeLeaving = 7,
    InformationTypeSelfLeaving = 8,
    InformationTypeAdminGranted = 9,
    InformationTypeAdminRemoved = 10,
    InformationTypeSelfAdminGranted = 11,
    InformationTypeSelfAdminRemoved = 12,
    InformationTypeSelfKicked = 13,
    InformationTypeGroupGone = 14
};

enum ChatType
{
    ChatTypeNone = 0,
    ChatTypeContact = 1,
    ChatTypeRoom = 2
};

// FIXME (boiko): I think this needs to be changed to a simple enum and not flags,
// as the statuses are mutually exclusive
enum AttachmentFlag
{
    AttachmentDownloaded = 0x01,
    AttachmentPending = 0x02,
    AttachmentError = 0x04
};

Q_DECLARE_FLAGS(AttachmentFlags, AttachmentFlag)

enum ParticipantState
{
    ParticipantStateRegular = 0,
    ParticipantStateLocalPending = 1,
    ParticipantStateRemotePending = 2
};

enum ParticipantRoles
{
    ParticipantRoleNone = 0,
    ParticipantRoleMember = 1,
    ParticipantRoleAdmin = 2
};

// Event writing results
enum EventWriteResult {
    EventWriteCreated,
    EventWriteModified,
    EventWriteError
};

// dbus service, object path and interface
static const char* DBusService = "com.canonical.HistoryService";
static const char* DBusObjectPath = "/com/canonical/HistoryService";
static const char* DBusInterface = "com.canonical.HistoryService";
static const char* ThreadViewInterface = "com.canonical.HistoryService.ThreadView";
static const char* EventViewInterface = "com.canonical.HistoryService.EventView";

// fields
static const char* FieldAccountId = "accountId";
static const char* FieldThreadId = "threadId";
static const char* FieldEventId = "eventId";
static const char* FieldType = "type";
static const char* FieldParticipants = "participants";
static const char* FieldParticipantIds = "participantIds";
static const char* FieldCount = "count";
static const char* FieldUnreadCount = "unreadCount";
static const char* FieldSenderId = "senderId";
static const char* FieldTimestamp = "timestamp";
static const char* FieldDate = "date";
static const char* FieldNewEvent = "newEvent";
static const char* FieldChatType = "chatType";
static const char* FieldChatRoomInfo = "chatRoomInfo";
static const char* FieldChatRoomJoined = "joined";
static const char* FieldChatRoomSelfRoles = "selfRoles";

// Chat Room Info Fields
static const char* FieldChatRoomName = "roomName";
static const char* FieldChatRoomServer = "server";
static const char* FieldChatRoomCreator = "creator";
static const char* FieldChatRoomCreationTimestamp = "creationTimestamp";
static const char* FieldChatRoomAnonymous = "anonymous";
static const char* FieldChatRoomInviteOnly = "inviteOnly";
static const char* FieldChatRoomParticipantLimit = "participantLimit";
static const char* FieldChatRoomModerated = "moderated";
static const char* FieldChatRoomTitle = "title";
static const char* FieldChatRoomDescription = "description";
static const char* FieldChatRoomPersistent = "persistent";
static const char* FieldChatRoomPrivate = "private";
static const char* FieldChatRoomPasswordProtected = "passwordProtected";
static const char* FieldChatRoomPassword = "password";
static const char* FieldChatRoomPasswordHint = "passwordHint";
static const char* FieldChatRoomCanUpdateConfiguration = "canUpdateConfiguration";

// thread fields
static const char* FieldLastEventId = "lastEventId";
static const char* FieldLastEventTimestamp = "lastEventTimestamp";
static const char* FieldGroupedThreads = "groupedThreads";

// text event fields
static const char* FieldMessage = "message";
static const char* FieldMessageType = "messageType";
static const char* FieldMessageStatus = "messageStatus";
static const char* FieldReadTimestamp = "readTimestamp";
static const char* FieldSubject = "subject";
static const char* FieldInformationType = "informationType";
static const char* FieldAttachments = "attachments";

// text attachment fields

static const char* FieldAttachmentId = "attachmentId";
static const char* FieldContentType = "contentType";
static const char* FieldFilePath = "filePath";
static const char* FieldStatus = "status";

// voice event fields
static const char* FieldMissed = "missed";
static const char* FieldDuration = "duration";
static const char* FieldRemoteParticipant = "remoteParticipant";

// sort stuff
static const char* FieldSortField = "sortField";
static const char* FieldSortOrder = "sortOrder";
static const char* FieldCaseSensitivity = "caseSensitivity";

// filter stuff
static const char* FieldFilterType = "filterType";
static const char* FieldFilterProperty = "filterProperty";
static const char* FieldFilterValue = "filterValue";
static const char* FieldMatchFlags = "matchFlags";
static const char* FieldFilters = "filters";
static const char* FieldGroupingProperty = "groupingProperty";

// contact matching stuff
static const char* FieldContactId = "contactId";
static const char* FieldAlias = "alias";
static const char* FieldAvatar = "avatar";
static const char* FieldIdentifier = "identifier";
static const char* FieldDetailProperties = "detailProperties";
static const char* FieldParticipantState = "state";
static const char* FieldParticipantRoles = "roles";

}

#endif // TYPES_H
